pkg/operator/upgradebackupcontroller/upgradebackupcontroller.go

```diff
@@ -0,0 +1,258 @@
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configv1informers "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdenvvar"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/etcd_assets"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// UpgradeBackupController responds to an upgrade request to 4.9 by attempting
+// to ensure the cluster is backed up. The CVO is expected to wait on a ceo
+// condition indicating successful backup before responding to the upgrade
+// request. This is intended to ensure that an upgrade from 4.8 to 4.9 does not
+// proceed without a recent backup to ensure restore a cluster that is not
+// healthy after upgrade to 4.9.
+
+const (
+	backupConditionType       = "RecentBackup"
+	clusterBackupVersionLabel = "cluster-backup-version"
+)
+
+type UpgradeBackupController struct {
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	podLister            corev1listers.PodLister
+	nodeLister           corev1listers.NodeLister
+	networkLister        configv1listers.NetworkLister
+	clusterVersionLister configv1listers.ClusterVersionLister
+	targetImagePullSpec  string
+	operatorVersion      string
+}
+
+func NewUpgradeBackupController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	clusterVersionInformer configv1informers.ClusterVersionInformer,
+	networkInformer configv1informers.NetworkInformer,
+	eventRecorder events.Recorder,
+	targetImagePullSpec string,
+	operatorVersion string,
+) factory.Controller {
+	c := &UpgradeBackupController{
+		operatorClient:       operatorClient,
+		operatorVersion:      operatorVersion,
+		kubeClient:           kubeClient,
+		podLister:            kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
+		nodeLister:           kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+		networkLister:        networkInformer.Lister(),
+		clusterVersionLister: clusterVersionInformer.Lister(),
+		targetImagePullSpec:  targetImagePullSpec,
+	}
+	return factory.New().ResyncEvery(10*time.Minute).WithInformers(
```

Comment:
Does this make the latency between the CVO asking for a backup, and etcd claiming the backup's completed 0-to-10 minutes for etcd to notice, and then a few minutes more for the backup itself?  Or do the informers allow the etcd operator to notice with an edge trigger, and this 10 minutes is... something else?

Reply:
Good question we have an informer on clusterVersion so any change to that resource would cause sync(). The 10 minute resync would be in addition to any change in informer resources. So assuming steady-state this controller would resync every 10m.

```
clusterVersionInformer.Informer(),
```

Reply:
The timing here should be very fast. I would expect the normal situation to be a few seconds (20 or less) from the time you update to backup complete.

Reply:
if it's sub-minute, I don't think there's any need for metrics coverage.  Might be interesting to emit events around this so we can spot-check a few in CI.  I dunno if there's an event emitter close by in the code that we could pipe in or not.